### PR TITLE
BUG: Add warning if rows have more columns than expected

### LINKED ISSUE #33037: Inconsistent behavior of `read_csv` when given an additional value on the first row of CSV file
#### Short description

Using the following invalid CSV:
```csv
col1_name,col2_name,col3_name
0,1,2,X
4,5,6,
6,7,8
```
It has 3 columns but the *first row* has 4 values and loading it should raise a `ParsingError`.

However, loading using:
```python
pd.read_csv("test.csv", index_col=False)
```
does not raise any exception and returns the following DataFrame:
```
   col1_name  col2_name  col3_name
0          0          1          2
1          4          5          6
2          6          7          8
```
Thus, it silently drops the additional value `X`.

#### Problem description

This will happening if the first row and the following ones are invalid.
If the first invalid row is not the first row, it will throw a `ParsingError` exception.
I.e. the following CSV produces the same results (silently dropping additional value):
```csv
col1_name,col2_name,col3_name
0,1,2,X
4,5,6,X
6,7,8
```
```csv
col1_name,col2_name,col3_name
0,1,2,X
4,5,6,X
6,7,8,X
```
```csv
col1_name,col2_name,col3_name
0,1,2,X
4,5,6
6,7,8,X
```
But this, as expected, throws an exception:
```csv
col1_name,col2_name,col3_name
0,1,2
4,5,6,X
6,7,8,X
```
Finally, if there are two additional values instead of a single one, it throws the following exception:
```
IndexError: list index out of range
```

Having a consistent behavior by throwing an exception in every of the previous cases would be enjoyable.
The fact that it is silent make it harder to validate CSV files.

#### Expected Output

Throw a `ParsingError` in the previous cases.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.7.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.15.0-91-generic
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.0.3
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 46.1.1.post20200323
Cython           : 0.29.15
pytest           : 5.4.1
hypothesis       : None
sphinx           : 2.4.4
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.1
IPython          : 7.13.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.1.3
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : 5.4.1
pyxlsb           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : None
</details>

#### Edits
- corrected copy-paste error in the returned DataFrame example (cf. [reply by @gfyoung](https://github.com/pandas-dev/pandas/issues/33037#issuecomment-604811382))

### PR DESCRIPTION
**Pull request by mproszewska**
- [x] closes #33037
- [x] tests added and passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] warning when read rows have more columns than expected

### CODE REVIEW

**Comment by mproszewska** [comment]
Multiple tests are not passing due to the new warning. 
Before fixing this, I want to ask if this solution is okay?

File: pandas/io/parsers.py [hunk 1]
```diff
@@ -2508,6 +2512,13 @@ def read(self, rows=None):
             content = content[1:]
 
         alldata = self._rows_to_cols(content)
+        if len(columns) != len(alldata) and notna(alldata[len(columns) :]).any():
+            warnings.warn(
+                "Expected {} columns instead of {}".format(len(columns), len(alldata)),
+                ParserWarning,
+                stacklevel=2,
+            )
+
```
Comment (gfyoung) [inline pandas/io/parsers.py:2521]: Abstract this into a helper function

File: pandas/io/parsers.py [hunk 2]
```diff
@@ -2508,6 +2512,13 @@ def read(self, rows=None):
             content = content[1:]
 
         alldata = self._rows_to_cols(content)
+        if len(columns) != len(alldata) and notna(alldata[len(columns) :]).any():
```
Comment (gfyoung) [inline pandas/io/parsers.py:2515]: Is there a reason we need the `notna` check?

**Comment by gfyoung** [comment]
@mproszewska : Sorry for the long wait here!  Overall solutions is on the right track.

File: pandas/io/parsers.py (hunk 2 above)
Comment (mproszewska) [reply pandas/io/parsers.py:2515]: In example mentioned in linked issue additional comma was added in one row. I assumed that additional commas are common and hence we might ignore them and don't raise a warning.
I'm using notna to check if data that won't be included contains only NaN values.

File: pandas/tests/io/parser/test_common.py [hunk 3]
```diff
@@ -2124,5 +2124,13 @@ def test_no_header_two_extra_columns(all_parsers):
...
+    with tm.assert_produces_warning(ParserWarning):
+        df = parser.read_csv(stream, header=None, names=column_names, index_col=False)
+        tm.assert_frame_equal(df, ref)
+
+
+def test_first_row_length(all_parsers):
+    stream = StringIO("col1,col2,col3\n0,1,2,X\n4,5,6\n6,7,8")
+    parser = all_parsers
+    with tm.assert_produces_warning(ParserWarning):
+        parser.read_csv(stream, index_col=False)
```
Comment (gfyoung) [inline pandas/tests/io/parser/test_common.py:2136]: Could you add the example from original issue (with the stray comma after the "6")

**Comment by mproszewska** [comment]
I have no idea how to check where this ipython directive error comes from.

**Comment by jbrockmendel** [comment]
can you rebase and ill take a look at the ipython thing

**Comment by mproszewska** [comment]
> can you rebase and ill take a look at the ipython thing

I rabased it

**Comment by jbrockmendel** [comment]
on the docbuild, it looks like the following is issuing a warning

```
data = "a,b,c\n4,apple,bat,\n8,orange,cow,"
pd.read_csv(StringIO(data), index_col=False)
```

under this PR, is issuing a warning here the correct thing to do?  If so, then an `:okwarning:` needs to go in io.rst on L756

**Comment by mproszewska** [comment]
> on the docbuild, it looks like the following is issuing a warning
> 
> ```
> data = "a,b,c\n4,apple,bat,\n8,orange,cow,"
> pd.read_csv(StringIO(data), index_col=False)
> ```
> 
> under this PR, is issuing a warning here the correct thing to do? If so, then an `:okwarning:` needs to go in io.rst on L756

I think so. First row has 3 values and the rest - 4. Where in in.rst should :okwarning: be added? maybe there's another way to do that. It shouldn't be a common warning.

**Comment by jreback** [comment]
conceptually this is ok. pls merge master and will re-look (and yes we would have to either fix the warnings or assert_produces_warning, though prob should fix the incorrect usages).

**Comment by jreback** [comment]
hmm this looks like overlapping with https://github.com/pandas-dev/pandas/pull/38587

**Comment by jreback** [comment]
closing in favor of https://github.com/pandas-dev/pandas/pull/38587

**Pull request closed without merging**